src/ray/raylet/scheduling/cluster_resource_data.cc

```diff
@@ -311,6 +311,20 @@ TaskResourceInstances NodeResourceInstances::GetAvailableResourceInstances() {
   return task_resources;
 };
 
+bool TaskRequest::IsEmpty() const {
```

Comment:
This and other new code bits are copied from the previous random actor placement PR.